Naor-Pinkas Base OT optimization with elliptic curves

Gugi264 opened the pull request:
Closes #375 

This PR optimizes the base Naor-Pinkas OT through the use of elliptic curves.
For this the curve p-256 is used (but can be changed easily).

The PR adds 3 new optional dependencies to the ot submodule.
1 For bouncy castle,
2 for the use of the commercial ECCelerate Library.
These dependencies are optional, the dependency and the corresponding files can be deleted if they are not wanted.

But at least one of those two is needed to use the optimization.

jonas-lj commented on tools/ot/src/main/java/dk/alexandra/fresco/tools/ot/base/BigIntElement.java:
```diff
@@ -0,0 +1,36 @@
+  }
+
+  @Override
+  public InterfaceNaorPinkasElement groupOp(InterfaceNaorPinkasElement other) {
```
I think this could be done more elegantly by adding a generic parameter to InterfaceNaorPinkasElement to avoid type casting

Gugi264 replied:
Now uses templates, no cast needed anymore

jonas-lj reviewed (changes requested):
Thanks for your PR! It looks great but we have a few comments - both on smaller issues and a few larger ones. My main concern is that the old OT seems to disappear - do you think it's possible to retain the original OT and add yours as a new one to optionally use?

Gugi264 commented on core/src/main/java/dk/alexandra/fresco/framework/util/ByteAndBitConverter.java:
```diff
@@ -72,6 +72,17 @@ public static BitSet intToBitSet(int i) {
     return res;
   }
 
+  public static String bytesToHex(byte[] bytes) {
```
Yes I removed this

Gugi264 commented on tools/ot/pom.xml:
```diff
@@ -21,5 +21,23 @@
       <artifactId>commitment</artifactId>
       <version>1.3.1-SNAPSHOT</version>
     </dependency>
+    <dependency>
```
I removed it in the newest commit

Gugi264 commented on suite/tinytables/src/main/java/dk/alexandra/fresco/suite/tinytables/ot/TinyTablesNaorPinkasOt.java:
```diff
@@ -1,45 +1,43 @@
+import dk.alexandra.fresco.tools.ot.base.AbstractNaorPinkasOT;
+import dk.alexandra.fresco.tools.ot.base.BouncyCastleNaorPinkas;
+
+public class TinyTablesNaorPinkasOt implements TinyTablesOt {
```
TinyTablesNaorPinkasOt uses stillt the old Naor Pinkas Implementation, just with the new name (BigIntNaorPinkas)

Gugi264 commented on suite/tinytables/src/main/java/dk/alexandra/fresco/suite/tinytables/ot/TinyTablesNaorPinkasOt.java:
```diff
@@ -1,45 +1,43 @@
-  }
-
-}
+package dk.alexandra.fresco.suite.tinytables.ot;
```
I think that was a problem with the line seperators, should be fixed now

Gugi264 commented on demos/common/src/main/java/dk/alexandra/fresco/demo/cli/CmdLineProtocolSuite.java:
```diff
@@ -229,8 +226,7 @@ private Drbg getDrbg(int myId, int prgSeedLength) {
       if (myId != otherId) {
-        DHParameterSpec dhSpec = DhParameters.getStaticDhParams();
-        Ot ot = new NaorPinkasOt(otherId, drbg, network, dhSpec);
+        AbstractNaorPinkasOT ot = new BigIntNaorPinkas(otherId, drbg, network);
```
BigIntNaorPinkas has the same functionality as the old NaorPinkasOt.
The name is just different to differentiate between different NaorPinkas implementations.

Gugi264 commented on tools/ot/src/test/java/dk/alexandra/fresco/tools/ot/base/TestFunctionalNaorPinkas.java:
```diff
@@ -49,10 +70,14 @@ public void shutdown() {
     try {
       Drbg rand = new AesCtrDrbg(HelperForTests.seedOne);
-      Ot otSender = new NaorPinkasOt(2, rand, network, staticParams);
+      Class clazz = this.testClass;
+      Constructor[] constructors = clazz.getConstructors();
+      Ot otSender = (AbstractNaorPinkasOT) constructors[0]
+          .newInstance(2, rand, network);
```
Used to get the constructor via reflection. Is not the most beautiful solution, but the best I could come up with without changing too much of the existing code

Gugi264 commented:
Thanks for the review!
I changed the files you highlighted.
 >  My main concern is that the old OT seems to disappear - do you think it's possible to retain the original OT and add yours as a new one to optionally use?
The old one doesn't disappear (at least not the functionality), I just refactored it and it is now just called BigIntNaorPinkas. If you don't want the  version using elliptic curves, just use that one.

jot2re reviewed (changes requested):
Looks good, but I think in general it is better to switch to Curve25519 as it is generally regarded as safer.

Gugi264 commented on tools/ot/src/main/java/dk/alexandra/fresco/tools/ot/base/BouncyCastleNaorPinkas.java:
```diff
@@ -0,0 +1,63 @@
+  /**
+   * The generator of the Diffie-Hellman group used in the OT.
+   */
+  private final org.bouncycastle.math.ec.ECPoint dhGenerator;
```
I refactored it so that only one type is used

Gugi264 commented on tools/ot/src/main/java/dk/alexandra/fresco/tools/ot/base/BouncyCastleNaorPinkas.java:
```diff
@@ -0,0 +1,63 @@
+  public BouncyCastleNaorPinkas(int otherId, Drbg randBit, Network network) {
+    super(otherId, randBit, network);
+    Security.addProvider(new BouncyCastleProvider());
+    X9ECParameters ecP = CustomNamedCurves.getByName("P-256");
```
Done

Gugi264 commented on core/src/test/java/dk/alexandra/fresco/framework/builder/numeric/field/MersennePrimeFieldDefinitionTest.java:
```diff
@@ -22,6 +23,21 @@ public void largeConstant() {
+  public void createElement() {
+      Long baseNumber = 2L;
+      MersennePrimeFieldDefinition fieldDefintion = new MersennePrimeFieldDefinition(4, 2);
+      MersennePrimeFieldElement elementFromString = (MersennePrimeFieldElement) fieldDefintion.createElement(baseNumber.toString());
```
No idea why that is, as highlighted from you, it is called

jot2re reviewed (approved):
Looks good! Thanks a lot for your contribution!

jot2re reviewed (changes requested):
Okay, I was a bit too quick approving. I see the build service has an issue building the project. It gi ved the following error `/home/runner/work/fresco/fresco/core/src/test/java/dk/alexandra/fresco/framework/builder/numeric/field/MersennePrimeFieldDefinitionTest.java:[36,39] error: cannot find symbol`
However, I did not have that issue when running the branch locally. 
Comparing it with what is shown here at GitHub it seems that there is an a weird whitespace symbol on this line.

Gugi264 commented:
I also didnt have the problem locally. I also dont see the whitespace problem, but I hope I fixed it.

Gugi264 commented:
I tried running the github actions on a different branch on my repository, and there I didn't get the error

jonas-lj commented:
I haven't run the code myself, but I think it's because the extractValue method on FieldElement was renamed to toBigInteger in september last year: https://github.com/aicis/fresco/commit/f128d39a2e2f090f497a0b6dc7309eec66db4f9e

jonas-lj commented:
> I tried running the github actions on a different branch on my repository, and there I didn't get the error See https://github.com/Gugi264/fresco/runs/5304055567?check_suite_focus=true

I haven't run the code myself, but I think it's because the extractValue method on FieldElement was renamed to toBigInteger in september last year: https://github.com/aicis/fresco/commit/f128d39a2e2f090f497a0b6dc7309eec66db4f9e

jonas-lj commented:
Looks good. Thanks for your contribution :)

jonas-lj commented:
Looks good. Thanks for your contribution :)
